GitHub Actions Sandbox

Tyler-g-hudson opened the pull request:
This PR contains three changes to the GitHub repository:

1. It implements workflow which contains a matrix of five GitHub Actions runs - Linux Latest Release, Linux Latest Debug, Linux Minimum Release, macOS Latest Release, and macOS Latest Debug. These each build the conda environment, build the repository, and run ctest on it with a varying set of OS, dependencies, and other options as well as excluding certain known problem tests.
2, 3. A ".github/conda_envs/" folder has been added to the repo which contains two environment files, one which contains the typical conda environment information and one of which describes a more restricted minimum requirements environment.

Tyler-g-hudson committed:
Initial Commit - Tyler
+ Created a GitHub Actions workflow
    This workflow is intended to try and work out a GitHub Actions workflow
    The intended function of the workflow is currently to install, build, and test the project on a Mac system
+ Added the environment.yaml file for use with conda

Tyler-g-hudson committed:
Fixed GHA line openers
~ Some "name:" lines in the GitHub Actions workflow were missing a "-"; resolved the problem

Tyler-g-hudson committed:
Ninja install in GHA workflow
+ Added a line to install ninja, also made conda update less verbose

Tyler-g-hudson committed:
Install FFTW
+ Added a line in the workflow configuration to also install FFTW due to an error on run
- Removed a line from the build command that was likely to cause problems

Tyler-g-hudson committed:
Resolved missing character

Tyler-g-hudson committed:
Added shells to some GHA tasks

Tyler-g-hudson committed:
Shell bash -l default, changed environment setup to use setup-miniconda

Tyler-g-hudson committed:
Small bugfix
~ Replaces "run" with "uses" to run outside GitHub Action

Tyler-g-hudson committed:
Matrix version
+ Now runs on both ubuntu and macOS
+ Attempted fix of macOS failure to install at cmake step

Tyler-g-hudson committed:
Minor typo fix

Tyler-g-hudson committed:
The Return Of -DCMAKE_INSTALL_PREFIX
+ Added REPO_PREFIX holding the position of the repo for macOS and Linux
+ Added -DCMAKE_INSTALL_PREFIX back into the cmake call under Configure CMAKE

Tyler-g-hudson committed:
Moved cd build to testing (ctest file was not found on run)

Tyler-g-hudson committed:
Fail-fast off
~ fail-fast set to off in order to permit full testing of macOS version, which runs quite a bit slower than ubuntu

Tyler-g-hudson committed:
Added export for DYLD_LIBRARY_PATH in Build Project step
+ Export was added as an attempt to fix a problem with python tests failing at this stage

Tyler-g-hudson committed:
Print tests, ctest exclude
+ Added some print tests to see what the environment variables are doing at the Test Project step
+ Added an exclude statement for a test that is expected to fail on any GHA system

Tyler-g-hudson committed:
Another attempt with environment variables
+ Added some new environment variables to the workflow to see if it can find Python (especially in MacOS)
+ Minor logic change in ctest

Tyler-g-hudson committed:
Minor bugfix
~ Forgot to put "conda" at the beginning of my conda env config vars set commands

Tyler-g-hudson committed:
Added environment variable preset step
+ Step performs environment variable setup
- Removed spurious uses of $REPO_PREFIX; ${GITHUB_WORKSPACE} already does this job
~ Small typo fix

Tyler-g-hudson committed:
(TEST) Ignore CXX tests, output on failure
~ Modified runner to output on failure to assist debugging macOS issue

Tyler-g-hudson committed:
Correct workspace prefix pathing
+ Added /install/prefix to the ${GITHUB_WORKSPACE} sections to point to the intended place

Tyler-g-hudson committed:
Debug removed, macOS test excludes re-added
- Removed debug code to print environment variables now I know they're correct
+ Added a variable for macos-latest that excludes raster and snaphu tests

Tyler-g-hudson committed:
Fixed ctest excludes
~ Modified how ctest exclusions work in the workflow file in order to permit more flexible exclusion of tests in ctest

Tyler-g-hudson committed:
Typo fix

Tyler-g-hudson committed:
Second attempt with environment variable

Tyler-g-hudson committed:
Tag-up suggestions
+ Added cancel-in-progress functionality with a concurrency group
~ Modified environment variables to point into the install file rather than the $CONDA environment
- Removed DCMAKE_CXX_FLAGS vestigial command from cmake config
~ Modified DCMAKE_INSTALL_PREFIX command to use new $ISCE3_PREFIX variable instead of what it has been doing
+ Added additional Install Project step to further separate and specify run steps

Tyler-g-hudson committed:
Concurrence test commit pt. 1
+ Added a single empty file to allow me to commit for the concurrence test

Tyler-g-hudson committed:
Concurrence test commit pt. 2
- Removed the concurrence-test file to allow a second commit

Tyler-g-hudson committed:
Switched to Mamba
- removed Python step as this should be handled by the Conda install
+ Added Mamba for faster building (hopefully)

Tyler-g-hudson committed:
Mamba, a different way
~ Removed miniconda setup, added micromamba setup
~ Had a good time doing it, too :)

Tyler-g-hudson committed:
Ubuntu minimum requirements run
+ Added extra specs field to the mamba setup step
+ Added a minimum requirements run on Ubuntu

Tyler-g-hudson committed:
Typo correction
+ Re-added runs-on statement that was accidentally removed

Tyler-g-hudson committed:
Modified run definitions
~ Modified the way the GHA ctest_exclude variables are set

Tyler-g-hudson committed:
Another try at a minimum requirements run
+ Changed default runner to Linux, runs macOS as an exception
~ Modified how runs are named

Tyler-g-hudson committed:
Moved ctest_exclude under os

Tyler-g-hudson committed:
Fixed indentation include -> os -> runner, deps

Tyler-g-hudson committed:
Minor fixes

Tyler-g-hudson committed:
De-indented items

Tyler-g-hudson committed:
More tab indentation checking

Tyler-g-hudson committed:
Typo fix

Tyler-g-hudson committed:
Moved all to include:

Tyler-g-hudson committed:
Attempt with adjusted minimum requirements

Tyler-g-hudson committed:
Minimum requirements build
+ added a minimum requirements environment yml file
~ moved environment files into their own folder
~ modified github workflow to use the yml files instead of a list of extra specs

Tyler-g-hudson committed:
Typo fix

Tyler-g-hudson committed:
Added GSL requirement
~ just to see what it does
~ Isn't this exciting?

Tyler-g-hudson committed:
Min reqs bugfix
~ --test-dir build doesn't seem to work on earlier versions, switched back to original method

Tyler-g-hudson committed:
Pin to Python 3.8 to check issues

Tyler-g-hudson committed:
Bump gdal for min reqs

Tyler-g-hudson committed:
Minimum requirements reverted, tests omitted (for now)
~ Reverted gdal back to 3.0; switching to 3.1 didn't fix anything
~ Removed several problem tests from the workflow:
	- removed ionosphere tests, these fail the minimum requirements tests for confusing reasons. Will address and add back in
	- removed test.python.pybind.unwrap.phass, this occasionally segfaults
	- removed test.python.pkg.isce3.io.background, this test has some logic may warrant changing due to the performance of lower-end systems

Tyler-g-hudson committed:
Attempt at DCMAKE_BUILD_TYPE variations
~ Changes to how the matrix is generated
~ Switched the DCMAKE_BUILD_TYPE option on cmake to use a variable

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+name: Build-And-Run Tests
+
+on: [push]
```
Can you add `pull_request` here as well?

Ideally, the CI build should be triggered whenever a PR is first created and whenever the PR is updated (by the PR author pushing changes to the PR branch). Right now, a CI build will only be triggered by the latter. (It didn't trigger when you created this PR, for example.)

```suggestion
on: [pull_request, push]
```

gmgunter commented on env/environment.yml:
```diff
@@ -0,0 +1,26 @@
+  - gmock>=1.10
+  - gtest>=1.10
+  - h5py>=3.0
+  - hdf5>=1.10.2,<1.13
```
Initially I set this upper bound on the HDF5 version, because my experience is that the HDF5 folks don't follow [semantic versioning](https://semver.org/) and tend to introduce a lot of breaking changes even in their patch releases. So when I'm showing people how to build isce3, I try to avoid building with a newer version of HDF5 than I personally have tested with, in case some new breaking changes have been introduced.

But it occurs to me now that we should remove this upper bound for purposes of CI testing for basically the same reason -- if a new breaking change is introduced in a recent HDF5 version, it would be good for the CI system to let us know early on so we can address it quickly.

```suggestion
  - hdf5>=1.10.2
```

gmgunter commented on env/minimum-reqs.yml:
```diff
@@ -0,0 +1,51 @@
+  # cmake: Current ver = 3.25
+  - cmake=3.18
+  - cxx-compiler
+  #####################
+  # EIGEN: Important
+  # Current ver: 3.4.0
```
I think we can remove the comments like this, since they're not providing much useful information now that we've got a working minimum environment.

The `Current ver: 3.4.0` is not super useful because it's specific to a particular platform at a particular point in time. This version could change without warning as new packages versions are added to conda.

The `EIGEN: Important` part was useful initially to indicate which packages we thought might be most likely to introduce breakages in isce3 upon version changes (so we could try harder to pin those at their exact minimum supported version and be a little more flexible about the other packages). Now that we've got a minimum-reqs environment, though, I'm not sure that we need to persist this info in a comment.

There's not really a strict distinction between an "Important" dependency and an unimportant one. This was just my intuition about which dependencies we should prioritize.

gmgunter commented on env/environment.yml:
```diff
@@ -0,0 +1,26 @@
+name: isce3-ci-sandbox
```
Let's just name it `isce3`

```suggestion
name: isce3
```

gmgunter commented on env/minimum-reqs.yml:
```diff
@@ -0,0 +1,51 @@
+name: isce3-ci-sandbox
```
```suggestion
name: isce3
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+        uses: mamba-org/provision-with-micromamba@main
+        with:
+          environment-file: ${{ matrix.deps.env-file }}
+          environment-name: isce3-ci-sandbox
```
```suggestion
          environment-name: isce3
```

gmgunter reviewed (commented):
Nice work @Tyler-g-hudson! Here are a few quick notes on things that I missed in our earlier offline discussions

gmgunter commented on env/environment.yml:
```diff
@@ -0,0 +1,26 @@
+name: isce3-ci-sandbox
```
I think I'd prefer that we don't create a new top-level directory to store these environment YAML files. My preference would be to try to keep the root directory of the repository as simple as possible and avoid adding too much clutter (although admittedly we're not currently doing a great job of that in isce3 :grin:).

Instead, I think it might be better to either put these in the `.github/` directory (perhaps in `.github/envs/`?) so that they live near the workflow that uses them, or perhaps put them somewhere in the `tools/`, where most of the existing CI scripts & resource files live.

I think I'm slightly more in favor of putting them in `.github/envs/` than in `tools/`, but I'm also slightly apprehensive that the GitHub folks might decide to standardize that directory name for something else at some point in the future (like they do with `.github/workflows/` and `.github/actions/`), which could possibly create issues.

Tyler-g-hudson replied:
Thoughts on using something less likely to be overwritten, like `.github/conda_envs/`?

gmgunter replied:
Yeah, good idea, `.github/conda_envs/` sounds good to me :+1:

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+jobs:
+  build_and_test:
+    env:
+      ALL_EXCLUDE: stage_dem|ionosphere|background|pybind.unwrap.phass
```
`ctest -E` uses a regex pattern to identify tests to exclude, so a`.` like this will match any character. If we want to match a literal `.` instead, we should escape the dot with a backslash, like this:

```suggestion
      ALL_EXCLUDE: stage_dem|ionosphere|background|pybind\.unwrap\.phass
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+            ctest_exclude: -E ".*($ALL_EXCLUDE)"
+          - runner: macos-latest
+            label: macOS
+            ctest_exclude: -E ".*($ALL_EXCLUDE|raster|unwrap.snaphu)"
```
```suggestion
            ctest_exclude: -E ".*($ALL_EXCLUDE|raster|unwrap\.snaphu)"
```

gmgunter replied:
There's actually a few different tests with "raster" in their name, and I think I only expect one of them to fail on macOS.

Let's see if we can make this pattern a little more specific in order to avoid unnecessarily excluding some tests

```suggestion
            ctest_exclude: -E ".*($ALL_EXCLUDE|io\.gdal\.raster|unwrap.snaphu)"
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+jobs:
+  build_and_test:
+    env:
+      ALL_EXCLUDE: stage_dem|ionosphere|background|pybind.unwrap.phass
```
Should all of these tests really be in `ALL_EXCLUDE`?

My impression was that the `background` and `pybind.unwrap.phass` test only failed on macOS.

And I thought the `ionosphere` tests were only failing on the Linux Minimum job.

If that's the case, let's see if we can move these excludes to the build matrix below so they're only excluded from specific jobs where they're expected to fail instead of being excluded from all workflows, as currently implemented.

Tyler-g-hudson replied:
My thinking here was that we were concerned about the nature of these tests generally, i.e. that they were written in such a way that they had greater potential to fail across a variety of platforms and dependencies (that is to say, if we were to add more tests to this CI workflow, all of the tests here had either a greater potential to fail or were considered otherwise flawed)

With the exception of stage_dem of course, which simply has a permissions problem and will be tested elsewhere.

If you'd like, I can move them though!

gmgunter replied:
Oh I see the reasoning now, thanks.

It's a fair point, but I'm thinking that it'd be better to try and maximize coverage of each job, and only add a test to `ALL_EXCLUDE` if it's been demonstrated to fail across multiple platforms and/or sets of dependencies. It doesn't cost much to move a test to `ALL_EXCLUDE` in the future. 

It seems like the `background` test is failing on macOS because we made some assumptions about runtime that aren't satisfied on the slower GHA macOS runners, and the `ionosphere` tests seem to be failing in the Linux Minimum job due to some issue with older versions of GDAL (though I'm not sure exactly what). So, at least in those two cases, we have some mechanistic reason to believe the tests should fail only in specific jobs.

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+          export ISCE3_PREFIX=${GITHUB_WORKSPACE}/install
+          conda env config vars set ISCE3_PREFIX=$ISCE3_PREFIX
+          conda env config vars set PYTHONPATH=$PYTHONPATH:$ISCE3_PREFIX/packages
+          if test -d "$ISCE3_PREFIX/lib64"; then
```
>`test -d "$ISCE3_PREFIX/lib64"`

I think this line isn't actually working the way we want it to, because we haven't yet installed isce3.

Our intention here is to check whether the isce3 libraries were installed in `$ISCE3_PREFIX/lib64` or `$ISCE3_PREFIX/lib` (because the install location differs on different linux distributions). We're doing this by testing whether the directory `$ISCE3_PREFIX/lib64` exists.

But we haven't yet installed isce3 at this point, so that directory won't exist yet in either case. Instead, we should perform this check after the installation step (but before the `ctest` step). At that point, the directory containing the installed libraries (either `lib` or `lib64`) should exist.

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+          conda env config vars set ISCE3_PREFIX=$ISCE3_PREFIX
+          conda env config vars set PYTHONPATH=$PYTHONPATH:$ISCE3_PREFIX/packages
+          if test -d "$ISCE3_PREFIX/lib64"; then
+            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64:$CONDA_PREFIX/bin/ld
+          else
+            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib:$CONDA_PREFIX/bin/ld
```
I don't think it should be necessary to add the `ld` linker itself to `LD_LIBRARY_PATH`

```suggestion
            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
          else
            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+          conda env config vars set ISCE3_PREFIX=$ISCE3_PREFIX
+          conda env config vars set PYTHONPATH=$PYTHONPATH:$ISCE3_PREFIX/packages
+          if test -d "$ISCE3_PREFIX/lib64"; then
+            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64:$CONDA_PREFIX/bin/ld
+          else
+            export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib:$CONDA_PREFIX/bin/ld
+          fi
+          conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH
```
Is it necessary to both `export LD_LIBRARY_PATH` \*and\* add `LD_LIBRARY_PATH` to the conda environment using `conda env config vars set`?

I think we could probably skip the first step and just add it to the conda environment

```suggestion
            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
          else
            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
          fi
```

Tyler-g-hudson replied:
The reason I'm doing these export-and-then-add-to-environment things is to prevent the runner from having to deactivate and reactivate the environment. `conda env config vars set` requires the environment to be deactivated and reactivated before the variable becomes available for use - so if I need the variable later (e.g. using $ISCE3_PREFIX to set $LD_LIBRARY_PATH and then using $LD_LIBRARY_PATH to set $DYLD_LIBRARY_PATH) I'll have to deactivate the environment and reactivate it which requires some extra steps and and a couple of extra seconds of runtime.

Using `export` first sets the environment variable and makes it available for use within this step and requiring no deactivation and reactivation, and then using `conda env config vars set` makes it available for use in the following steps of the workflow. I've found this to be a faster and simpler way of doing the job.

Another alternative is also for me to just set these variables without referencing each other at all, which I believe would require an additional if/else statement and may make the script a little slower to modify but would delete these export/conda env config vars set steps:

```Shell
conda env config vars set ISCE3_PREFIX=${GITHUB_WORKSPACE}/install
conda env config vars set PYTHONPATH=$PYTHONPATH:${GITHUB_WORKSPACE}/install/packages
if test -d "${GITHUB_WORKSPACE}/install/lib64"; then
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib64
  if [ "$RUNNER_OS" == "macOS" ]; then
    conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib64
  fi
else
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib
  if [ "$RUNNER_OS" == "macOS" ]; then
    conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:${GITHUB_WORKSPACE}/install/lib
  fi
fi
```

Tyler-g-hudson replied:
Scratch this! With the above comment about moving LD_LIBRARY_PATH setting to after `cmake --build build --target install` has been run, I don't end up having much of a reason to do these export statements anyway.

gmgunter replied:
Oh right gotcha -- you needed to export `LD_LIBRARY_PATH` as well because you were using it to populate `DYLD_LIBRARY_PATH` as well in that same workflow step. Sorry, I forgot about that part when I made the initial comment but it makes sense now.

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,109 @@
+      # Preset environment variables in the conda environment for later steps
+      - name: Environment Variables Preset
+        run: |
+          export ISCE3_PREFIX=${GITHUB_WORKSPACE}/install
```
Just a small stylistic nitpick. Elsewhere, when we're evaluating environment variables, we just use `$VAR` (without curly braces) instead of `${VAR}`. 

This is the only exception to that here AFAICT, so let's ditch the braces for aesthetic reasons. They shouldn't be necessary in this case.

```suggestion
          export ISCE3_PREFIX=$GITHUB_WORKSPACE/install
```

gmgunter reviewed (commented):
Few more suggestions :slightly_smiling_face:

Tyler-g-hudson committed:
PR changes pt. 1
- Removed extra comments from minimum-reqs.yml
- Removed upper bound on hdf5 from environment.yml
~ Changed exclusion test names in build-and-run.yaml from containing . to \. for more accurate and bug-free regular expressions, among a few other small changes
~ Moved LD_LIBRARY_PATH and DYLD_LIBRARY_PATH setting in build-and-run.yaml to after the repository has been built
~ Changed isce3-ci-sandbox naming convention to just isce3 for PR

Tyler-g-hudson committed:
Merge branch 'github-actions-sandbox' of git@github.com:Tyler-g-hudson/isce3-ci-sandbox.git into github-actions-sandbox

# Conflicts:
#	.github/workflows/build-and-run.yml
#	env/minimum-reqs.yml

Tyler-g-hudson committed:
Bug correction for macOS
~ Moved the setting of LD_LIBRARY_PATH and DYLD_LIBRARY_PATH to their own step, to ensure that these variables are available to the tester when it runs.

Signed-off-by: Tyler Hudson <tyler.g.hudson@jpl.nasa.gov>

Tyler-g-hudson committed:
PR corrections pt. 2
~ Moved items in envs folder to .github/conda_envs/
~ Moved some tests from ALL_EXCLUDE to specific testing runs where we believe they are likely or expected to fail.

gmgunter commented:
I'm not sure why we're not seeing the CI checks here on the PR page. I see it on @Tyler-g-hudson's fork here - https://github.com/Tyler-g-hudson/isce3-ci-sandbox/actions - but normally we'd see the pass/fail results on the PR itself.

Maybe the `.github/workflows/*.yml` file needs to already exist when the PR is created for this to work? Or maybe it's because the initial PR creation didn't trigger a CI workflow?

Either way, I suppose it's probably not a big deal. We can just check that the workflows on the fork are passing for now. After this is merged, I expect that we'll see CI checks on subsequent PRs, but we can verify that when we get to that point.

rtburns-jpl commented:
Yeah I think that's a security feature, so we can inspect the workflow runs before they run under the organization. Once it's merged they'll run here as you expect.

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+    steps:
+      # Typical github repo checkout step
+      - name: Github Repo Checkout
+        uses: actions/checkout@v2
```
Let's see if we can [upgrade to `checkout@v3`](https://github.com/actions/checkout#whats-new). I see a few warnings on our [CI runs](https://github.com/Tyler-g-hudson/isce3-ci-sandbox/actions/runs/4167096208) currently that seem to be related to using the older version:

![Screenshot from 2023-02-14 21-21-40](https://user-images.githubusercontent.com/12984092/218939843-f0bd35cc-ee27-484f-84c5-1737ea2e164d.png)

```suggestion
        uses: actions/checkout@v3
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+      # Set the LD_LIBRARY_PATH and DYLD_LIBRARY_PATH variables
+      - name: Set Link Library Environment Variables
+        run: |
+          if test -d "$ISCE3_PREFIX/lib64"; then
+            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
+            if [ "$RUNNER_OS" == "macOS" ]; then
+              conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
+            fi
+          else
+            conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
+            if [ "$RUNNER_OS" == "macOS" ]; then
+              conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$ISCE3_PREFIX/lib
+            fi
+          fi
```
According to [this](https://stackoverflow.com/questions/3146274/is-it-ok-to-use-dyld-library-path-on-mac-os-x-and-whats-the-dynamic-library-s), `DYLD_LIBRARY_PATH` on macOS behaves the same as `LD_LIBRARY_PATH` on Linux. So, on macOS, I think we \*only\* need to define `DYLD_LIBRARY_PATH` -- not both `LD_LIBRARY_PATH` \*and\* `DYLD_LIBRARY_PATH`.

Also, the `lib64` vs `lib` directory name ambiguity is really only an issue on Linux. According to [this](https://stackoverflow.com/a/67720256), `lib64` is used only on non-debian-based 64-bit Linux systems. On macOS, the library install directory will always be called `lib` -- never `lib64`.

With this in mind, I think we could try to simplify this logic a bit more. WDYT about this approach?

```
# Append to DYLD_LIBRARY_PATH on macOS or to LD_LIBRARY_PATH on Linux.
# Depending on the Linux flavor, the install dir name could be 'lib64' or 'lib'.
if [ "$RUNNER_OS" == "macOS" ]; then
  conda env config vars set DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$ISCE3_PREFIX/lib
elif test -d "$ISCE3_PREFIX/lib64"; then
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib64
else
  conda env config vars set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ISCE3_PREFIX/lib
fi
```

Tyler-g-hudson replied:
Looks good! I always love cutting a layer out of my conditional logic. I'll give it a try and see if it runs!

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass
+
+
+    defaults:
```
Another very minor pedantic comment --

In general in this file, you use a single blank line to separate blocks of code. But here you're using 2 blank lines. I think it would look a little cleaner if we tried to consistently use the same number of empty lines to space things out

```suggestion
      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass

    defaults:
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+          conda env config vars set PYTHONPATH=$PYTHONPATH:$GITHUB_WORKSPACE/install/packages
+
+      
+      # Get cmake command configured with relevant build information
```
Same here

```suggestion
          conda env config vars set PYTHONPATH=$PYTHONPATH:$GITHUB_WORKSPACE/install/packages

      # Get cmake command configured with relevant build information
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+jobs:
+  build_and_test:
+    env:
+      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass
```
Last but not least, let's add a comment here to explain why we're excluding these particular tests, because otherwise I will 100% forget why we did things this way :grin:

```suggestion
      # stage_dem requires AWS permissions to access the S3 bucket containing the DEM data.
      # pybind.unwrap.phass occasionally segfaults on GHA runners.
      ALL_EXCLUDE: stage_dem|pybind\.unwrap\.phass
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+            ctest_exclude: -E ".*($ALL_EXCLUDE)"
+          - runner: macos-latest
+            label: macOS
+            ctest_exclude: -E ".*($ALL_EXCLUDE|io\.gdal\.raster|unwrap\.snaphu|background)"
```
```suggestion
            # io.gdal.raster and unwrap.snaphu tests fail due to known unresolved issues
            # with GDAL memory mapping on macOS.
            # background reader/writer test sometimes fails on macOS due to slow runtimes.
            ctest_exclude: -E ".*($ALL_EXCLUDE|io\.gdal\.raster|unwrap\.snaphu|background)"
```

gmgunter commented on .github/workflows/build-and-run.yml:
```diff
@@ -0,0 +1,116 @@
+          - os:
+              label: Linux
+              runner: ubuntu-latest
+              ctest_exclude: -E ".*($ALL_EXCLUDE|ionosphere)"
```
```suggestion
              # ionosphere tests fail on Linux Minimum due to apparent issues with older GDAL versions
              ctest_exclude: -E ".*($ALL_EXCLUDE|ionosphere)"
```

Tyler-g-hudson committed:
Update .github/workflows/build-and-run.yml

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Tyler-g-hudson committed:
PR suggestions pt. 3
+ Added comments describing reasoning for text exclusions on various platforms
- Removed extra newlines between some parts of the workflow script
~ Modified logic of assigning LD_LIBRARY_PATH and DYLD_LIBRARY_PATH to fit the internal logic of macOS and Linux systems

gmgunter reviewed (dismissed):
LGTM, thanks!

Tyler-g-hudson committed:
Obsessive Compulsive Disorder
~ There is now only one Link Library Environment Variable being set

gmgunter commented:
@rtburns-jpl we talked over this PR together in our devops meeting a few weeks ago. Do you have any other comments that you want to make here? If not, can you go ahead and approve when you get a chance?

After this is merged, I was planning to ask @Tyler-g-hudson to look into replacing the environment.yml files with requirements.txt files, since we discussed wanting to minimize the number of redundant places in the codebase where isce3's requirements are enumerated.

rtburns-jpl reviewed (approved):
Nope nothing to add, all looking great to me.
